### PR TITLE
Add config_only_args arg to alf.configurable to prevent local config changes

### PR DESCRIPTION
**Pull request by QuantuMope**
This PR adds the `config_only_args` arg to the `alf.configurable` decorator as well as a unit test.

The functionality can be described as:
1. Any arg specified in this list can only be changed via `alf.config` and will otherwise raise a `ValueError`. 
2. Any entry in `config_only_args` cannot also be found in `blacklist`, otherwise a `ValueError` is raised.

This will allow users to eliminate possible mismatches in parameters that may arise from local config changes using `partial()`.

### CODE REVIEW

File: alf/config_util.py [hunk 1]
```diff
@@ -557,13 +561,21 @@ def _wrapper(*args, **kwargs):
...
+                    )
+
         return fn(*args, *unspecified_positional_args, **kwargs,
                   **unspecified_kw_args)
 
     return _wrapper
 
 
-def _decorate(fn_or_cls, name, whitelist, blacklist):
+def _decorate(fn_or_cls, name, whitelist, blacklist, guard_kwargs):
```
Comment (hnyu) [inline alf/config_util.py]: It's not actually 'kwargs'? It's better that we also have this guard for positional arg?

File: alf/config_util.py [hunk 2]
```diff
@@ -557,13 +561,21 @@ def _wrapper(*args, **kwargs):
                     unspecified_kw_args[name] = config.get_value()
                 config.set_used()
 
+        for key in guard_kwargs:
+            if key in kwargs and key in configs:
```
Comment (hnyu) [inline alf/config_util.py]: We can just directly assert that any key in guard_kwargs shouldn't appear in kwargs or args? (i.e., it has to be in `configs`)

**Review by hnyu — CHANGES_REQUESTED** [review]

File: alf/config_util.py [hunk 3]
```diff
@@ -576,6 +588,10 @@ def _decorate(fn_or_cls, name, whitelist, blacklist):
         blacklist (list[str]): A blacklisted set of kwargs that should not be
             configurable. All other kwargs will be configurable. Only one of
             ``whitelist` or ``blacklist`` should be specified.
+        guard_kwargs (list[str]): list of kwargs that should be guarded. In other
```
Comment (emailweixu) [inline alf/config_util.py]: The name `guard_kwargs` does really convey the meaning. How about  `config_only_args`?

File: alf/config_util.py (hunk 2 above)
Comment (emailweixu) [reply alf/config_util.py]: +1

Comment (QuantuMope) [reply alf/config_util.py]: Done.

File: alf/config_util.py (hunk 1 above)
Comment (QuantuMope) [reply alf/config_util.py]: I've expanded the feature for positional args as well.

File: alf/config_util.py (hunk 3 above)
Comment (QuantuMope) [reply alf/config_util.py]: Good idea. Changed.

**Review by QuantuMope — COMMENTED** [review]
Addressed comments. PTAL.

File: alf/config_util.py [hunk 4]
```diff
@@ -527,6 +527,10 @@ def _make_wrapper(fn, configs, signature, has_self):
         has_self (bool): whether the first argument is expected to be self but
             signature does not contains parameter for self. This should be True
             if fn is __init__() function of a class.
+        config_only_args (list[str]): list of args that should be guarded. In other
+            words, their values can only be set / changed globally via alf.config().
+            This protects against the case where local changes via partial() can
+            cause unintended side effects.
```
Comment (hnyu) [inline alf/config_util.py]: This also protects the case where a value is specified by the code when calling the function?

File: alf/config_util.py [hunk 5]
```diff
@@ -557,13 +562,21 @@ def _wrapper(*args, **kwargs):
                     unspecified_kw_args[name] = config.get_value()
                 config.set_used()
 
+        for config_only_arg in config_only_args:
+            if config_only_arg in set_positional_args or config_only_arg in kwargs.keys(
```
Comment (hnyu) [inline alf/config_util.py]: Can be directly `config_only_arg in kwargs`

**Review by hnyu — COMMENTED** [review]
Only some minors comments

File: alf/config_util.py (hunk 4 above)
Comment (QuantuMope) [reply alf/config_util.py]: Right. Updated comments to mention this case as well.

File: alf/config_util.py (hunk 5 above)
Comment (QuantuMope) [reply alf/config_util.py]: Changed.

**Review by QuantuMope — COMMENTED** [review]
Addressed comments. PTAL.

**Review by hnyu — APPROVED** [review]

**Pull request merged**